clang/lib/CIR/CodeGen/CIRGenCXX.cpp

```diff
@@ -310,47 +310,109 @@ void CIRGenModule::codegenGlobalInitCxxStructor(const VarDecl *D,
-    lexScope.setAsGlobalInit();
+  llvm_unreachable("NYI");
+}
 
-    builder.setInsertionPointToStart(block);
-    Address DeclAddr(getAddrOfGlobalVar(D), getASTContext().getDeclAlign(D));
-    buildDeclInit(CGF, D, DeclAddr);
-    builder.setInsertionPointToEnd(block);
-    builder.create<mlir::cir::YieldOp>(Addr->getLoc());
+void CIRGenModule::buildCXXGlobalVarDeclInit(const VarDecl *varDecl,
+                                             mlir::cir::GlobalOp addr,
+                                             bool performInit) {
+  const Expr *init = varDecl->getInit();
+  QualType ty = varDecl->getType();
+
+  // TODO: handle address space
+  // The address space of a static local variable (DeclPtr) may be different
+  // from the address space of the "this" argument of the constructor. In that
+  // case, we need an addrspacecast before calling the constructor.
+  //
+  // struct StructWithCtor {
+  //   __device__ StructWithCtor() {...}
+  // };
+  // __device__ void foo() {
+  //   __shared__ StructWithCtor s;
+  //   ...
+  // }
+  //
+  // For example, in the above CUDA code, the static local variable s has a
+  // "shared" address space qualifier, but the constructor of StructWithCtor
+  // expects "this" in the "generic" address space.
+  assert(!MissingFeatures::addressSpace());
+
+  if (getLangOpts().OpenMP && !getLangOpts().OpenMPSimd &&
+      varDecl->hasAttr<OMPThreadPrivateDeclAttr>()) {
+    llvm_unreachable("NYI");
   }
 
-  if (isCstStorage) {
-    // TODO: this leads to a missing feature in the moment, probably also need a
-    // LexicalScope to be inserted here.
-    buildDeclInvariant(CGF, D);
-  } else {
-    // If not constant storage we'll emit this regardless of NeedsDtor value.
+  assert(varDecl && " Expected a global declaration!");
+  CIRGenFunction cgf{*this, builder, true};
+  CurCGF = &cgf;
+  CurCGF->CurFn = addr;
+
+  CIRGenFunction::SourceLocRAIIObject fnLoc{cgf,
+                                            getLoc(varDecl->getLocation())};
+
+  addr.setAstAttr(
+      mlir::cir::ASTVarDeclAttr::get(builder.getContext(), varDecl));
+
+  if (ty->isReferenceType()) {
     mlir::OpBuilder::InsertionGuard guard(builder);
-    auto block = builder.createBlock(&Addr.getDtorRegion());
-    CIRGenFunction::LexicalScope lexScope{*CurCGF, Addr.getLoc(),
+    auto *block = builder.createBlock(&addr.getCtorRegion());
+    CIRGenFunction::LexicalScope lexScope{*CurCGF, addr.getLoc(),
                                           builder.getInsertionBlock()};
     lexScope.setAsGlobalInit();
-
     builder.setInsertionPointToStart(block);
-    buildDeclDestroy(CGF, D);
+    auto getGlobal = builder.createGetGlobal(addr);
+
+    Address declAddr(getGlobal, getGlobal.getType(),
+                     getASTContext().getDeclAlign(varDecl));
+    assert(performInit && "cannot have constant initializer which needs "
+                          "destruction for reference");
+    RValue rv = cgf.buildReferenceBindingToExpr(init);
```

Comment:
@bcardosolopes Here is the problem mentioned -- this is a reference to a temporary here and thus the AST type is `ExprWithCleanup`. So this builds something like

```
 %0 = "cir.get_global"() <{name = @a}> : () -> !cir.ptr<!cir.ptr<!ty_A>>
    "cir.scope"() ({
      %1 = "cir.get_global"() <{name = @_ZGR1a_}> : () -> !cir.ptr<!ty_A>
      "cir.yield"() : () -> ()
    }) : () -> ()
    "cir.store"(%1, %0) : (!cir.ptr<!ty_A>, !cir.ptr<!cir.ptr<!ty_A>>) -> ()
```

So we need a way to get `%1` out of the `scope`.

Reply:
Ideally we should make the scope return the rvalue (which we do sometimes already), but it'd require changing other cases unrelated to this PR. I'll create an issue to track that and will look into it. For now just need to make sure that the store is created in the same basic block as the underlying LValue, this can be done by using the rvalue origin to set an insertion point. Example:

```
// Store the result close to the rvalue production.
// TODO(cir): might be better to change ExprWithCleanupsClass to (a) return the rvalue via the scope, (b) incorporate scope information into LValue instead.
{
  mlir::OpBuilder::InsertionGuard guard(builder);
  mlir::Operation *rvalDefOp = rv.getScalarVal()->getDefiningOp();
  if (rvalDefOp && rvalDefOp->getBlock()) {
     mlir::Block *rvalSrcBlock = rvalDefOp->getBlock();
     if (!rvalSrcBlock->empty() && isa<YieldOp>(rvalSrcBlock->back()) {
      auto yield = cast<YieldOp>(rvalSrcBlock->back());
      builder.setInsertionPoint(yield);
    }
  }
  cgf.buildStoreOfScalar(rv.getScalarVal(), declAddr, false, ty);
}
```